golem-rib/src/interpreter/rib_interpreter.rs

```diff
@@ -231,9 +277,138 @@ mod internal {
+                Some(result)
+            }
+            RibInterpreterResult::Val(TypeAnnotatedValue::Bool(bool)) => Some(bool),
+            RibInterpreterResult::Val(_) => None,
+            RibInterpreterResult::Unit => None,
+            RibInterpreterResult::Sink(_, _) => None,
+        };
+
+        if !predicate_bool.ok_or("Expecting a value that can be converted to bool".to_string())? {
+            instruction_stack.move_to(&instruction_id).ok_or(format!(
+                "Internal error: Failed to move to the instruction at {}",
+                instruction_id.index
+            ))?;
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn run_list_to_iterator_instruction(
+        interpreter_stack: &mut InterpreterStack,
+    ) -> Result<(), String> {
+        let existing_sink = interpreter_stack
+            .pop()
+            .ok_or("A sink has to exist before any iteration")?;
+
+        if !existing_sink.is_sink() {
+            return Err("Expecting a sink before iteration".to_string());
+        }
+
+        // Loading the result in the previous evaluation
+        if let Some(RibInterpreterResult::Val(TypeAnnotatedValue::List(items))) =
+            interpreter_stack.pop()
+        {
+            let iter = items
+                .values
+                .into_iter()
+                .map(|x| x.clone().type_annotated_value.unwrap());
+
+            interpreter_stack.push(existing_sink);
+            interpreter_stack.push(RibInterpreterResult::Iterator(Box::new(iter)));
+
+            Ok(())
+        } else {
+            Err("Expected a List on the stack for ListToIterator".to_string())
+        }
+    }
+
+    pub(crate) fn run_create_sink_instruction(
+        interpreter_stack: &mut InterpreterStack,
+        analysed_type: &AnalysedType,
+    ) -> Result<(), String> {
+        let analysed_type = match analysed_type {
+            AnalysedType::List(type_list) => type_list.clone().inner,
+            _ => return Err("Expecting a list type to create sink".to_string()),
+        };
+        interpreter_stack.create_sink(analysed_type.deref());
+        Ok(())
+    }
+
+    pub(crate) fn run_advance_iterator_instruction(
+        interpreter_stack: &mut InterpreterStack,
+    ) -> Result<(), String> {
+        let mut possible_iterator = interpreter_stack
+            .pop()
+            .ok_or("Failed to advance the iterator".to_string())?;
+
+        match &mut possible_iterator {
+            RibInterpreterResult::Iterator(iter) => {
+                if let Some(type_annotated_value) = iter.next() {
+                    interpreter_stack.push(possible_iterator);
+                    interpreter_stack.push(RibInterpreterResult::Val(type_annotated_value));
+                    Ok(())
+                } else {
+                    Err("Internal Error: Iterator has no more items".to_string())
+                }
+            }
+            _ => Err("Internal Error: Expected an Iterator on the stack".to_string()),
```

Comment:
This will never happen, however, hesitant to assume here and `unwrap`.

Reply:
Please note many of the internal errors`Err...` you see shouldn't happen, however, they are explicit mainly for two reasons

1. Hard to comprehend for another person if unwraps are safe or not.
2. Easier code maintainability, were can make logical mistakes and test case failures is easier to debug